sycl/doc/extensions/proposed/sycl_ext_oneapi_device_image_backend_content.asciidoc

```diff
@@ -0,0 +1,234 @@
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension adds a mechanism to obtain the raw backend content of the device
+images that are in a kernel bundle.
+The format of this content is implementation-defined, so applications that make
+use of this extension are normally not portable to other implementations of
+SYCL.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_DEVICE_IMAGE_BACKEND_CONTENT` to one of the values defined in
+the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New member functions in the `device_image` class
+
+This extension adds the following member functions to the `device_image` class.
+
+[source,c++]
+----
+namespace sycl {
+
+template <bundle_state State>
+class device_image {
+ public:
+  std::vector<std::byte> ext_oneapi_get_backend_content() const;
+  std::span<std::byte> ext_oneapi_get_backend_content_view() const;
+
+  /*...*/
+};
+
+} // namespace sycl
+----
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+std::vector<std::byte> ext_oneapi_get_backend_content() const;
+----
+!====
+
+_Constraints:_ Available only when `State` is `bundle_state::executable`.
+
+_Returns:_ A copy of the raw backend content for this device image.
+The format of this data is implementation-defined.
+See below for a description of this format on {dpcpp}.
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+std::span<std::byte> ext_oneapi_get_content_backend_view() const;
+----
+!====
+
+Available only when the compiler is {cpp}20 or higher.
+
+_Constraints:_ Available only when `State` is `bundle_state::executable`.
+
+_Returns:_ A view of the raw backend content for this device image.
+The data behind this view has the same lifetime as the `kernel_bundle` that
+contains this `device_image`.
+
+'''
+
+
+== Device image format for {dpcpp}
+
+This section is non-normative and applies only to the {dpcpp} implementation.
+The format of the data returned by
+`device_image::ext_oneapi_get_backend_content` and
+`device_image::ext_oneapi_get_backend_content_view` depends on the backend of the
+kernel bundle that contains the device image.
```

Comment:
Is there any way for the developer to query the backend associated with a `device_image`?  I might have missed it, but I can't find a `get_backend()` function in the SYCL 2020 `device_image` synopsis.

I don't know if that's an oversight, or if this extension should add an `ext_oneapi_get_backend` function.

Reply:
The `kernel_bundle` class has `get_backend`.  I figured that was enough since a `device_image` always comes from some `kernel_bundle`.

Reply:
Would you mind adding it as an open issue?  I agree that it being in `kernel_bundle` makes it less of a pressing issue, but this means a library function working on a `device_image` always needs to take a `kernel_bundle` or a `backend` as an argument if it wants to be able to call these functions.  Adding `get_backend()` to `device_image` would make things simpler for developers using this extension.

Reply:
I added `ext_oneapi_get_backend` in 236c32ebdf18ea1910a73d73860c0b24fb23be25.  I also changed the lifetime of the data returned by `ext_oneapi_get_content_backend_view` to make it easier for applications that use `device_image` without also having a copy of the containing `kernel_bundle` object.